content/interoperability.tex

```diff
@@ -0,0 +1,189 @@
+advice to \openshmem library users and developers that may improve the portability
+and performance of hybrid programs, and definition of an OpenSHMEM
+API that queries the interoperability features provided by an \openshmem library.
+
+
+\section{MPI Interoperability}
+
+\openshmem and MPI are two commonly used parallel programming models for
+distributed-memory systems. The user can choose to utilize both models in the same program
+to efficiently and easily support various communication patterns.
+
+A vendor may implement the \openshmem and MPI libraries in different ways. For
+instance, one may implement both \openshmem and MPI as standalone libraries,
+each of which allocates and initializes fully isolated communication
+resources.
+Another common approach
+is to implement both \openshmem and MPI interfaces within the
+same software system in order to share a communication resource when possible.
+
+To improve interoperability and portability in \openshmem + MPI hybrid
+programming, we clarify the relevant semantics in the following subsections.
+
+
+\subsection{Initialization}
+In order to ensure that a hybrid program can be portably performed with different vendor
+implementations, the \openshmem environment of the program must be initialized by
+a call to \FUNC{shmem\_init} or \FUNC{shmem\_init\_thread} and be finalized by
+a call to \FUNC{shmem\_finalize}; the MPI environment of the program must be initialized
+by a call to \FUNC{MPI\_Init} or \FUNC{MPI\_Init\_thread} and be finalized by a
+call to \FUNC{MPI\_Finalize}.
+
+\apiimpnotes{
+Portable implementations of OpenSHMEM and MPI must ensure that the initialization
+calls can be made in an arbitrary order within a program; the same rule also
+applies to the finalization calls. A software runtime that utilizes a shared
+communication resource for \openshmem and MPI communication may maintain an
+internal reference counter in order to ensure that the shared resource is
+initialized only once and thus no shared resource is released until the last
+finalization call is made.
+}
+
+
+\subsection{Dynamic Process Creation}
+\label{subsec:interoperability:mpmd}
+
+MPI defines a dynamic process model that allows creation of processes after
+an MPI application has started (e.g., by calling \FUNC{MPI\_Comm\_spawn}) and
+connection to independent processes (e.g., through \FUNC{MPI\_Comm\_accept}
+and \FUNC{MPI\_Comm\_connect})
+and provides a mechanism to establish communication
+between the newly created processes and the existing MPI application (see
+MPI standard version 3.1, Chapter 10).
+Unlike MPI, \openshmem starts all processes at once and requires all PEs to
+collectively allocate and initialize resources (e.g., symmetric heap) used by
+the \openshmem library before any other \openshmem routine may
+be called. \openshmem does not support communication with dynamically created
+or connected processes. In such a scenario, MPI must be used to communicate
+with these processes.
```

Comment:
I am actually a little cautious in making the possible communication approaches more restricted, because users may also use other approaches such as socket. We only want to prohibit OpenSHMEM in such scenarios. How about we say?
```In such a scenario, MPI can be used to communicate with these processes.```